### PR TITLE
update postgres output module

### PR DESCRIPTION
**Pull request by GrayTShirt**
Fixes issue #1824, invalidates #1800, implements #895, #648, #818

### CODE REVIEW

**Comment by rgerhards** [comment]
Thx for the patch set. The Solaris buildbot slave seems to have an unrelated issue. I will investigate and either re-start or report back if something is really with the patch.

**Comment by GrayTShirt** [comment]
this PR also resolves #818

**Comment by GrayTShirt** [comment]
Quick question; what's the difference between CODEqueryEtryPt_STD_OMODTX_QUERIES and CODEqueryEtryPt_TXIF_OMOD_QUERIES and generally which one is preferred?

**Comment by GrayTShirt** [comment]
For prepared statements, I'm stuck trying to figure out the number of params a SQL statement should have. For instance, given a particular template, how many variables will it have? Additionally is that something I can get before `doAction`?

**Comment by rgerhards** [comment]
> For prepared statements, I'm stuck trying to figure out the number of params a SQL statement should have. For instance, given a particular template, how many variables will it have? Additionally is that something I can get before doAction?

The whole interface is modeled after string passing. There *was* an array-interface that provided what you look for. However, the only user was omoracle, which was contributed and abandoned. So I have removed that interface a couple of years ago. I could look into reviving it, but it's unfortunately not done in an instant :-(

I would suggest that we go ahead with this PR as-is and then discuss how we can work towards an even better one. What do you think? Along the same lines, would you be willing to lend me a helping hand actually activating the tests during CI runs? I probably could use some advise on postgres packages and commands -- it's very long since I last used them...

**Comment by rgerhards** [comment]
Note: the travis failure is caused by this: https://bugs.llvm.org/show_bug.cgi?id=35572

I have disabled that static analyzer for now in current master (buildbot still has it!). If you rebase, that error will go away.

**Comment by GrayTShirt** [comment]
Sure, I can ensure we run tests on travis, I suggest we use the postgres provided apt repository. How do you feel about adding another build slot?

Too bad about the array interface. I'm still interested in an array/structured interface even if it's lots of work.

**Comment by rgerhards** [comment]
> Sure, I can ensure we run tests on travis, I suggest we use the postgres provided apt repository. How do you feel about adding another build slot?

I see you already added commits (Y) I wouldn't mind another build slot, but I think it is not vitally necessary. I need to refactor Travis scripts in any case soon:
1. you really notice they have grown and been rewoven over time, it's no longer straightforward
1. we will soon hit the 50 minute single-build timeout, so I need to distribute over more VMs (I am very grateful that Travis permits us to run such an extended test suite on their systems!)

> Too bad about the array interface. I'm still interested in an array/structured interface even if it's lots of work.

I  need to check what needs to be touched, but I fear it is a lot. IIRC the interface was functional the last time in the v6 or v7 timeframe, and many things have changed since then. Most importantly, I fear that we need to do quite a bit of core changes. Given the current todo, this sounds a bit challenging :-(

If you would like to have a shot at it, we can see how we could work together on it. But I think I would at least need to get you decently started, else it probably is too frustrating (due to the many fine things to be taken into account).

**Comment by rgerhards** [comment]
oh... in any case, I think we should merge **this** PR once the testbench runs. It really, really is very valuable! Thx for the great work.

**Comment by GrayTShirt** [comment]
My pleasure, I'm happy to help

**Comment by GrayTShirt** [comment]
There appears to be some backwards compatibility issues and a small memory leak. I'll continue debugging.

**Comment by rgerhards** [comment]
> not really, other than grouping it near other diag.sh commands, and the only command before it is SQL prep

@GrayTShirt The reason it should be the first is that it does post-run cleanup plus it provides the test heading. In theory, future versions of the test harness may file if init is not the first statement. Highly suggest moving it up.

**Comment by GrayTShirt** [comment]
Yea I realized that putting the test heading first might be a good idea :)

**Comment by GrayTShirt** [comment]
Ok, the backwards compatibility has been fixed, but there still is a memory leak in the new configuration syntax without a defined template.

**Comment by GrayTShirt** [comment]
I've added some more memory checks to figure out what the problem is, but I'm have trouble.

**Comment by GrayTShirt** [comment]
When I run the valgrind tests locally I fail all of them with 36000 bytes still reachable. But I'm not getting the 150000 bytes that travis is getting.

**Comment by rgerhards** [comment]
Which test do you refer to? Just do that I can find it quicker. Can you
also post a link to a relevant Travis log/run?

Sent from phone, thus brief.

**Comment by GrayTShirt** [comment]
https://travis-ci.org/rsyslog/rsyslog/jobs/313672627
Test: pgsql-basic-cnf6-vg.sh

Also on phone.

**Comment by rgerhards** [comment]
@GrayTShirt did you notice that the test was valgrind-clean but had a problem with the output data? From the bottom of its test log:
```
rsyslogd run exited with 0     --> valgrind is ok
scanf error in index i=0          --> expected output data seems not to be present
sequence error detected
```
The difference in remaining assigned memory can be due to --enable-valgrind option, if set differently (if set, it doesn't unload modules). It can also be due different libraries etc. Valgrind is used in the testbench so that any remaining allocs that are not reported as errors are 99.99% sure irrelevant.

**Comment by GrayTShirt** [comment]
Aye yi yi!

**Comment by GrayTShirt** [comment]
Shall I rebase into a single commit?

**Comment by rgerhards** [comment]
I think that's a great idea!

File: tests/pgsql-basic-cnf6-vg.sh [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+	action(type="ompgsql" server="127.0.0.1"
+		db="syslogtest" user="postgres" pass="testbench")
+}'
+. $srcdir/diag.sh startup-vg
+. $srcdir/diag.sh injectmsg  0 5000
+. $srcdir/diag.sh shutdown-when-empty
+. $srcdir/diag.sh wait-shutdown-vg
+. $srcdir/diag.sh check-exit-vg
+
+# we actually put the message in the SysLogTag field, so we know it doesn't use the default
```
Comment (rgerhards) [inline tests/pgsql-basic-cnf6-vg.sh]: I guess this comment is not fully correct ;-)

Comment (rgerhards) [reply tests/pgsql-basic-cnf6-vg.sh]: If I am right, a ```git commit --amend``` should do for the fix ;-)

**Comment by GrayTShirt** [comment]
I deleted that confusing comment :)

**Comment by GrayTShirt** [comment]
this PR will also resolve #895 and #648

**Comment by rgerhards** [comment]
I updated the initial description. Let me know (or update) if I missed some issues that it also closes ;-)

**Comment by GrayTShirt** [comment]
Those are all the open issues and pull-requests I found that relate to this change, good to go!

**Comment by rgerhards** [comment]
@GrayTShirt great work! Thanks a ton, this will be very useful for rsyslog users :-) I'll merge and close this PR now, feel free to open a new issue tracker for the transactional interface. We can then check what needs to be done to make it happen!

**Comment by rgerhards** [comment]
oh... I forgot: it would be great if you could update the doc as well: https://github.com/rsyslog/rsyslog-doc/blob/master/source/configuration/modules/ompgsql.rst

**Comment by GrayTShirt** [comment]
Hahaha! I'm already working on it!

**Pull request merged**